tests/test_sftp.py

```diff
@@ -0,0 +1,341 @@
+        from io import BytesIO
+
+        if "r" in mode:
+            return BytesIO(self.files[remote_file])
+        else:
+
+            def write(data):
+                self.files[remote_file] = data
+
+            return type("MockFile", (), {"write": write})
+
+    def close(self):
+        self.closed = True
+
+
+class MockSSHClient:
+    def __init__(self):
+        self.connected = False
+        self.sftp_client = MockSFTPClient()
+        self.closed = False
+
+    def set_missing_host_key_policy(self, policy):
+        pass
+
+    def connect(self, hostname, port, username, password=None, pkey=None):
+        self.connected = True
+        self.hostname = hostname
+        self.port = port
+        self.username = username
+        self.password = password
+        self.pkey = pkey
+
+    def open_sftp(self):
+        return self.sftp_client
+
+    def close(self):
+        self.closed = True
+
+
+# ============== Tests for SFTPDelivery ==============
+
+
+def test_sftp_delivery_valid_initialization(sftp_delivery):
+    assert sftp_delivery.host == "sftp.example.com"
+    assert sftp_delivery.port == 22
+    assert sftp_delivery.username == "user"
+    assert sftp_delivery.password == "password"
+    assert sftp_delivery.private_key is None
+
+
+def test_sftp_delivery_invalid_host():
+    with pytest.raises(TypeError):
+        SFTPDelivery(host=123, username="user", password="password")
+    with pytest.raises(ValueError):
+        SFTPDelivery(host="   ", username="user", password="password")
+
+
+def test_sftp_delivery_invalid_port():
+    with pytest.raises(TypeError):
+        SFTPDelivery(
+            host="sftp.example.com", username="user", password="password", port="22"
+        )
+    with pytest.raises(ValueError):
+        SFTPDelivery(
+            host="sftp.example.com", username="user", password="password", port=70000
+        )
+
+
+def test_sftp_delivery_invalid_username():
+    with pytest.raises(TypeError):
+        SFTPDelivery(host="sftp.example.com", username=123, password="password")
+    with pytest.raises(ValueError):
+        SFTPDelivery(host="sftp.example.com", username="", password="password")
+
+
+def test_sftp_delivery_invalid_password():
+    with pytest.raises(TypeError):
+        SFTPDelivery(host="sftp.example.com", username="user", password=123)
+    with pytest.raises(ValueError):
+        SFTPDelivery(host="sftp.example.com", username="user", password="")
+
+
+def test_sftp_delivery_invalid_private_key():
+    with pytest.raises(TypeError):
+        SFTPDelivery(
+            host="sftp.example.com",
+            username="user",
+            password="password",
+            private_key=123,
+        )
+    with pytest.raises(ValueError):
+        SFTPDelivery(
+            host="sftp.example.com",
+            username="user",
+            password="password",
+            private_key="   ",
+        )
+
+
+def test_sftp_delivery_connect(monkeypatch, sftp_delivery):
```

Comment:
This test should check that `__establish_ssh_connection` and `open_sftp` methods have been called.